### PR TITLE
Improve horizontal fusion of contiguous

### PR DESCRIPTION
**Pull request by pfultz2**

### CODE REVIEW

**Comment by causten** [comment]
@turneram can you take a look at these CI failures

File: test/simplify_reshapes_test.cpp [hunk 1]
```diff
@@ -1275,4 +1275,82 @@ TEST_CASE(transpose_slice_single_transpose)
...
+        auto transpose2 = m1.add_instruction(
+            migraphx::make_op("transpose", {{"permutation", {0, 1, 3, 2}}}), slice2);
+        auto slice3 = m1.add_instruction(
+            migraphx::make_op("slice", {{"axes", {1}}, {"starts", {24}}, {"ends", {36}}}),
+            transpose);
+        m1.add_return({slice1, transpose2, slice3});
+    }
+    auto output_shapes = m1.get_output_shapes();
+    run_pass(m1);
+    EXPECT(m1.get_output_shapes() == output_shapes);
```
Comment (umangyadav) [inline test/simplify_reshapes_test.cpp:1332]: CI is failing on this check, looks like we may need to update logic for  calculating strides for the `slice`.

File: src/simplify_reshapes.cpp [hunk 2]
```diff
@@ -664,9 +667,94 @@ struct find_slice_transpose
...
+        // Make transpose
+        std::transform(perm.begin(), perm.end(), perm.begin(), [&](auto i) {
+            if(i > preaxis)
+                return i + 1;
+            return i;
+        });
+        perm.insert(perm.begin(), preaxis + 1);
+        auto transpose =
+            m.insert_instruction(ins, make_op("transpose", {{"permutation", perm}}), unsqueeze);
+        // Slice and sqeeze
```
Comment (umangyadav) [inline src/simplify_reshapes.cpp:740]:
```suggestion
        // Slice and squeeze
```

File: src/simplify_reshapes.cpp [hunk 3]
```diff
@@ -664,9 +667,94 @@ struct find_slice_transpose
...
+        // TODO: Handle multiple axes
+        if(sdistance.size() != 1)
+            return;
+        auto axis = slice.axes.front();
+        // Skip if axis would be packed
+        if(std::all_of(ins->get_shape().lens().begin(),
+                       ins->get_shape().lens().begin() + axis,
+                       [](auto x) { return x == 1; }))
+            return;
+        // Compute axis before transpose to use for unsqueeeze
```
Comment (umangyadav) [inline src/simplify_reshapes.cpp:725]:
```suggestion
        // Compute axis before transpose to use for unsqueeze
```

**Review by umangyadav — CHANGES_REQUESTED** [review]

File: test/simplify_reshapes_test.cpp (hunk 1 above)
Comment (pfultz2) [reply test/simplify_reshapes_test.cpp:1332]: We cant change the strides to `slice` otherwise it will be incorrect. Perhaps this should just check the lens.

File: test/include/test.hpp [hunk 4]
```diff
@@ -136,6 +128,17 @@ inline auto operator<<(Stream& s, const Range& v) -> decltype(stream_range(s, v.
...
 
+template <class Stream, class Iterator>
+inline Stream& stream_range(Stream& s, Iterator start, Iterator last)
+{
+    if(start != last)
+    {
+        s << *start;
+        std::for_each(std::next(start), last, [&](auto&& x) { s << ", " << x; });
+    }
+    return s;
```
Comment (umangyadav) [inline test/include/test.hpp:139]: Was this change necessary ?

**Review by umangyadav — APPROVED** [review]

**Review by turneram — APPROVED** [review]
